### PR TITLE
Removed Memory._id

### PR DESCRIPTION
**Pull request by KrzysztofCwalina**
We want to get to baseline performance first before adding pooling checks/diagnostics. This change:

- makes Memory<T> smaller and so passing it by value and construction cheaper
- makes get_Span, TryGetPointer, TryGetArray cheaper as they won’t have to pass the ID and do ID checks
- makes diagnosing use after return bugs harder

#### Before (with ID)
```
                   Method |       Mean |    StdDev |     Median | Scaled | Scaled-StdDev |        RPS | Allocated |
------------------------- |----------- |---------- |----------- |------- |-------------- |----------- |---------- |
           ParsePlaintext |  2.0345 us | 0.0546 us |  2.0456 us |   1.00 |          0.00 | 491,511.89 |     105 B |
  ParsePipelinedPlaintext |  1.8487 us | 0.0524 us |  1.8607 us |   0.91 |          0.03 | 540,916.50 |     105 B |
          ParseLiveAspNet |  9.5836 us | 0.2583 us |  9.5493 us |   4.71 |          0.18 | 104,345.41 |   1.11 kB |
 ParsePipelinedLiveAspNet | 10.7971 us | 0.3896 us | 10.7426 us |   5.31 |          0.23 |  92,617.14 |   1.11 kB |
             ParseUnicode | 14.8633 us | 0.3654 us | 15.0547 us |   7.31 |          0.26 |  67,279.60 |   1.94 kB |
    ParseUnicodePipelined | 16.6655 us | 0.3759 us | 16.6712 us |   8.20 |          0.28 |  60,004.12 |   1.94 kB |
```

#### After (no ID)

```
                   Method |       Mean |    StdDev |     Median | Scaled | Scaled-StdDev |        RPS | Allocated |
------------------------- |----------- |---------- |----------- |------- |-------------- |----------- |---------- |
           ParsePlaintext |  1.9474 us | 0.0695 us |  1.9281 us |   1.00 |          0.00 | 513,500.72 |     105 B |
  ParsePipelinedPlaintext |  1.6893 us | 0.0513 us |  1.6757 us |   0.87 |          0.04 | 591,974.34 |     105 B |
          ParseLiveAspNet |  9.1254 us | 0.2620 us |  9.2597 us |   4.69 |          0.21 | 109,583.94 |    1.1 kB |
 ParsePipelinedLiveAspNet | 10.1330 us | 0.1856 us | 10.2064 us |   5.21 |          0.21 |  98,687.41 |    1.1 kB |
             ParseUnicode | 14.3673 us | 0.5455 us | 14.4018 us |   7.39 |          0.38 |  69,602.74 |   1.94 kB |
    ParseUnicodePipelined | 16.1511 us | 0.3173 us | 16.1940 us |   8.30 |          0.33 |  61,915.45 |   1.94 kB |
```

### CODE REVIEW

**Comment by KrzysztofCwalina** [comment]
cc: @shiftylogic, @ahsonkhan, @joshfree, @pakrym

File: src/System.Slices/System/Buffers/OwnedMemory.cs [hunk 1]
```diff
@@ -140,39 +140,34 @@ protected virtual void OnZeroReferences()
         #endregion
 
         #region Used by Memory<T>
-        void IKnown.AddReference(long id)
+        void IKnown.AddReference()
         {
             AddReference();
             try {
-                VerifyId(id);
             } catch (ObjectDisposedException e) {
```
Comment (ahsonkhan) [inline src/System.Slices/System/Buffers/OwnedMemory.cs]: Remove the try/catch block?

**Comment by ahsonkhan** [comment]
Remove the VerifyId method?
https://github.com/dotnet/corefxlab/blob/master/src/System.Slices/System/Buffers/OwnedMemory.cs#L213

**Comment by davidfowl** [comment]
You need to remove a bunch of tests for use after free.

**Comment by joshfree** [comment]
Would it make sense instead of removing _id completely to put it under #if DEBUG for now?

**Comment by davidfowl** [comment]
Nah, nuke it.

**Comment by KrzysztofCwalina** [comment]
@joshfree, I was trying to, but it's all over signatures. The source would get very hard to read. I can add it back later in a couple of hours.

File: src/System.Slices/System/Buffers/OwnedMemory.cs (hunk 1 above)
Comment (KrzysztofCwalina) [inline src/System.Slices/System/Buffers/OwnedMemory.cs]: Thanks. I removed.

**Comment by KrzysztofCwalina** [comment]
@mjp4, I was trying to add you as a reviewer, but it does not seem to work.

**Comment by ahsonkhan** [comment]
@mjp4, I was trying to add you as a reviewer, but it does not seem to work.

You mean @mjp41?

**Pull request merged**